### PR TITLE
feat(storage): rebuild s3/local dedupe index when switching dedupe status

### LINKED ISSUE #565: dedupe: configuration may switch from disabled to enabled
**Is your feature request related to a problem? Please describe.**

Consider dedupe configuration which is initially off, then the user enables it.

**Describe the solution you'd like**

Ideally, would like to build/rebuild the dedupe lookup index.

### PR DESCRIPTION
**Pull request by peusebiu**
closes #565

Signed-off-by: Petu Eusebiu <peusebiu@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/controller.go [hunk 1]
```diff
@@ -620,13 +625,16 @@ func (c *Controller) Shutdown() {
...
-	taskScheduler.RunScheduler(reloadCtx)
+	taskScheduler.RunScheduler(reloadCtx, scheduler.RateLimit)
 
 	// Enable running garbage-collect periodically for DefaultStore
 	if c.Config.Storage.GC && c.Config.Storage.GCInterval != 0 {
 		c.StoreController.DefaultStore.RunGCPeriodically(c.Config.Storage.GCInterval, taskScheduler)
 	}
 
+	// Enable running rebuild dedupe index
+	c.StoreController.DefaultStore.RunRebuildDedupeIndex(time.Duration(0), taskScheduler, c.Config.Storage.Dedupe)
```
Comment (andaaron) [inline pkg/api/controller.go:636]: Are these durations correct? In all RunRebuildDedupeIndex in this file.

File: pkg/storage/s3/s3.go [hunk 2]
```diff
@@ -35,7 +35,8 @@ import (
 )
 
 const (
-	CacheDBName = "s3_cache"
+	CacheDBName    = "s3_cache"
+	LastDedupePath = "dedupedstatus.txt"
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/storage/s3/s3.go:39]: yeah, forgot to remove it, it's from the last implementation

File: pkg/api/controller.go (hunk 1 above)
Comment (peusebiu) [reply pkg/api/controller.go:636]: time.Duration(0) means that scheduler is not periodic, just one time thing.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -85,9 +85,9 @@ func NewScheduler(logC log.Logger) *Scheduler {
 	}
 }
 
-const rateLimit = 5 * time.Second
+const RateLimit = 5 * time.Second
 
-func (scheduler *Scheduler) RunScheduler(ctx context.Context) {
+func (scheduler *Scheduler) RunScheduler(ctx context.Context, rateLimit time.Duration) {
```
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/scheduler/scheduler.go:90]: But why would this be a parameter of RunScheduler(). Do we call this function multiple times with different values? Why not set it when initializing the scheduler?

[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/scheduler/scheduler.go:90]: makes sense, I'll move it to the constructor

**Review by andaaron — APPROVED** [review]

**Review by andaaron — APPROVED** [review]
LGTM

**Review by andaaron — APPROVED** [review]

**Comment by rchincha** [comment]
resolve conflicts and re-push pls

**Comment by peusebiu** [comment]
> resolve conflicts and re-push pls

done

File: pkg/storage/s3/s3_test.go [hunk 4]
```diff
@@ -752,8 +768,8 @@ func TestNegativeCasesObjectsStorage(t *testing.T) {
 			controller := api.NewController(conf)
 			So(controller, ShouldNotBeNil)
 
-			err = controller.InitImageStore(context.Background())
-			So(err, ShouldBeNil)
+			// err = controller.InitImageStore(context.Background())
```
Comment (andaaron) [inline pkg/storage/s3/s3_test.go:771]: Is this supposed to be commented out?

**Review by andaaron — APPROVED** [review]
I'm not familiar with the actual dedupe implementation inside storage, but overall looks good.

File: pkg/storage/s3/s3_test.go (hunk 4 above)
Comment (peusebiu) [reply pkg/storage/s3/s3_test.go:771]: nop, thanks :D

File: .github/workflows/nightly.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+name: 'Storage dedupe/restore blobs task - nightly'
```
Comment (rchincha) [inline .github/workflows/nightly.yaml:1]: Just call this "Nightly jobs"

This dedupe job is just one of them. Likely will add more later.

**Comment by rchincha** [comment]
minor nitpicks, otherwise lgtm

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**